proposals/0000-bundle-with-type-synonym.rst

```diff
@@ -0,0 +1,138 @@
+.. proposal-number:: Leave blank. This will be filled in when the proposal is
+                     accepted.
+
+.. trac-ticket:: Leave blank. This will eventually be filled with the Trac
+                 ticket number which will track the progress of the
+                 implementation of the feature.
+
+.. implemented:: Leave blank. This will be filled in with the first GHC version which
+                 implements the described feature.
+
+.. highlight:: haskell
+
+Bundle pattern synonyms with a type synonym
+===========================================
+
+With `AssociatingSynonyms <https://ghc.haskell.org/trac/ghc/wiki/PatternSynonyms/AssociatingSynonyms>`_
+it is possible to associate pattern synonyms with a datatype,
+enabling them to be imported just like a data constructor using the ``Type(..)`` syntax.
+Pattern synonyms can also be used to offer a simplified view on a type,
+in which case it would be more natural to associate them with a type synonym
+for the corresponding simplified type,
+e.g., ``State`` instead of ``StateT`` for the state monad.
+Therefore, support for associating pattern synonyms with type aliases seems desirable.
+
+Motivation
+----------
+
+Consider the following motivating example, which is a stripped-down version of
+`mtl <https://hackage.haskell.org/package/mtl>`_'s ``Identity`` and ``State`` monads.
+
+::
+
+ {-# LANGUAGE ViewPatterns #-}
+ {-# LANGUAGE PatternSynonyms #-}
+ {-# LANGUAGE ScopedTypeVariables #-}
+ 
+ module State (
+     Identity(..),
+     State, pattern State, runState
+ ) where
+ 
+ import Control.Monad
+ 
+ newtype Identity a = Identity { runIdentity :: a }
+ 
+ newtype StateT s m a = StateT { runStateT :: s -> m (s, a) }
+ 
+ type State s a = StateT s Identity a
+ 
+ pattern State { runState } <- ((runIdentity .) . runStateT -> runState)
```

Comment:
This appears to be impossible at the moment. The left-hand side of the pattern does not allow a type signature, and adding a type signature to the right-hand side,

    pattern State { runState } <- ((runIdentity .) . runStateT -> runState) :: State s a

does not silence the warning. I will fix one of the name-shadowing warnings.